cpp/src/arrow/flight/sql/server_test.cc

```diff
@@ -0,0 +1,783 @@
+        const auto& list_item = dynamic_cast<const Int32Scalar&>(*list_item_scalar).value;
+        ASSERT_EQ(expected_int32_list[j], list_item);
+      }
+    }
+  }
+
+  explicit SqlInfoDenseUnionValidator(const DenseUnionScalar& data) : data(data) {}
+
+  SqlInfoDenseUnionValidator(const SqlInfoDenseUnionValidator&) = delete;
+  SqlInfoDenseUnionValidator(SqlInfoDenseUnionValidator&&) = delete;
+  SqlInfoDenseUnionValidator& operator=(const SqlInfoDenseUnionValidator&) = delete;
+};
+
+std::unique_ptr<TestServer> server;
+std::unique_ptr<FlightSqlClient> sql_client;
+
+class TestFlightSqlServer : public ::testing::Environment {
+ protected:
+  void SetUp() override {
+    server.reset(new TestServer("flight_sql_test_server"));
+    server->Start();
+    for (int i = 0; i < 100; i++) {
+      std::this_thread::sleep_for(std::chrono::milliseconds(10));
+      if (server->IsRunning()) {
+        break;
+      }
+    }
+    ASSERT_TRUE(server->IsRunning());
+
+    std::stringstream ss;
+    ss << "grpc://localhost:" << server->port();
+    std::string uri = ss.str();
+
+    std::unique_ptr<FlightClient> client;
+    Location location;
+    ASSERT_OK(Location::Parse(uri, &location));
+    ASSERT_OK(FlightClient::Connect(location, &client));
+
+    sql_client.reset(new FlightSqlClient(std::move(client)));
+  }
+
+  void TearDown() override {
+    sql_client.reset();
+    server->Stop();
+    server.reset();
+  }
+};
+
+TEST(TestFlightSqlServer, TestCommandStatementQuery) {
+  ASSERT_OK_AND_ASSIGN(auto flight_info,
+                       sql_client->Execute({}, "SELECT * FROM intTable"));
+
+  ASSERT_OK_AND_ASSIGN(auto stream,
+                       sql_client->DoGet({}, flight_info->endpoints()[0].ticket));
+
+  std::shared_ptr<Table> table;
+  ASSERT_OK(stream->ReadAll(&table));
+
+  const std::shared_ptr<Schema>& expected_schema =
+      arrow::schema({arrow::field("id", int64()), arrow::field("keyName", utf8()),
+                     arrow::field("value", int64()), arrow::field("foreignId", int64())});
+
+  std::shared_ptr<Array> id_array;
+  std::shared_ptr<Array> keyname_array;
+  std::shared_ptr<Array> value_array;
+  std::shared_ptr<Array> foreignId_array;
+  ArrayFromVector<Int64Type>({1, 2, 3}, &id_array);
+  ArrayFromVector<StringType, std::string>({"one", "zero", "negative one"},
+                                           &keyname_array);
+  ArrayFromVector<Int64Type>({1, 0, -1}, &value_array);
+  ArrayFromVector<Int64Type>({1, 1, 1}, &foreignId_array);
+
+  const std::shared_ptr<Table>& expected_table = Table::Make(
+      expected_schema, {id_array, keyname_array, value_array, foreignId_array});
+
+  AssertTablesEqual(*expected_table, *table);
+}
+
+TEST(TestFlightSqlServer, TestCommandGetTables) {
+  std::vector<std::string> table_types;
+  ASSERT_OK_AND_ASSIGN(
+      auto flight_info,
+      sql_client->GetTables({}, nullptr, nullptr, nullptr, false, table_types));
```

Comment:
This is not very readable, would recommend making the parameter names explicit, such as:
```c++
sql_client->GetTables(/*options=*/{}, /*catalog=*/nullptr, ...